### PR TITLE
Fix backward pass on GPU in PyTorch interface

### PR DESCRIPTION
**Pull request by glassnotes**
**Context:** A user has [reported an error](https://discuss.pennylane.ai/t/transfer-learning-error/1134/3) through the quantum transfer learning demo; when running on the GPU, the line
```
 vjp = dy.view(1, -1) @ ctx.jacobian.apply(ctx, *ctx.saved_tensors)
```
in `torch.py` yields:
```
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking arugment for argument mat2 in method wrapper_mm)
```

The reason for the error was discovered to be that the result of `ctx.jacobian.apply()`was on the CPU, even when the tensors it acted on were on the GPU.

**Description of the Change:** Updated the `backward` method in the torch interface to split the above operation into two parts, and check the device of the results. If at least one of them was on the GPU, the other is moved to the GPU if it was not already there.

Note: I needed to use torch 1.9 to test this, due to the issue described [here](https://github.com/allenai/allennlp/issues/5064).

**Benefits:** Running on the GPU with the torch interface should now work. 

**Possible Drawbacks:** While this fix enables the demo to be run on the GPU, we cannot fully test it.

**Related GitHub Issues:** I believe this is the same issue reported in #1290, so merging this should enable us to close that.

### CODE REVIEW

File: pennylane/interfaces/torch.py
Comment (josh146) [inline pennylane/interfaces/torch.py:201]: 🙌 thanks for fixing this @glassnotes!

Quick question, will this need to be fixed up on line 141 as well?

Comment (glassnotes) [reply pennylane/interfaces/torch.py:201]: Oh that's a really good question. When I ran the example, this was the section that actually threw the error. I've tested on both the transfer learning demo, and the small example in the docs #1225 and both work with the fix. Do you know of a code example that would lead to the earlier part being called?

Comment (josh146) [reply pennylane/interfaces/torch.py:201]: The other backward method is called for Hessian computations. That is, if you have

```python
hess = torch.autograd.functional.hessian(cost_fn, params)
```

**Review by albi3ro — COMMENTED** [review]
Before I approve, I'd like to know if we get the same errors on calculating the Hessian, as Josh mentioned.

I know this is a fairly small change, but I assume we still need a changelog for it?

**Comment by glassnotes** [comment]
> Before I approve, I'd like to know if we get the same errors on calculating the Hessian, as Josh mentioned.
> 
> I know this is a fairly small change, but I assume we still need a changelog for it?

Changelog updated!

Computing the Hessian went smoothly - I didn't have to update anything. It looks like there is some logic already in that section to keep things on the GPU (I'm genuinely confused why this doesn't happen for the backward pass though, I wonder if the actual issue is within the `.jacobian.apply` function).

File: pennylane/interfaces/torch.py
Comment (josh146) [inline pennylane/interfaces/torch.py:199]: It occurs to me that we could factor this out as a utility function, so that this logic can be re-used where needed in the future:

```python
def match_gpu_device(tensors):
    """Ensure that tensors are placed on the same device.

    This function checks whether any tensor within the input list of
    tensors is on a GPU. If this is the case, all tensors are moved
    to the location of the *first CUDA tensor in the list*.

    If no tensors are on a GPU, the tensor list is returned unchanged.

    Args:
        tensors (sequence[torch.Tensor]): list of input tensors
    """
    device = None

    for t in tensors:
        if t.is_cuda:
            device = t.get_device()
            break

    if device is None:
        return

    for t in tensors:
        if t.get_device() is not device:
            torch.as_tensor(t, device=device)
```

**Review by josh146 — APPROVED** [review]
Thanks @glassnotes! Looks good on my end. Approving with the caveat that I am unable to test it locally as I do not have a GPU, so recommend getting approval from someone that does.

**Comment by mlxd** [comment]
Just a quick comment, I have successfully run the [tutorial](https://pennylane.ai/qml/demos/tutorial_quantum_transfer_learning.html) on my device (GTX 1060) running ArchLinux, CUDA 11.3, Driver v465.31, PyTorch 1.9.0+CUDA11.1 using this branch. PyTorch 1.8.1 fails with cuBlas errors.

Output as below, along with the prediction windows:
```
% python ./tutorial_quantum_transfer_learning.py
Downloading: "https://download.pytorch.org/models/resnet18-f37072fd.pth" to /home/mlxd/.cache/torch/hub/checkpoints/resnet18-f37072fd.pth
100.0%
Training started:
/tmp/pl_gpu/pyenv/lib/python3.9/site-packages/torch/nn/functional.py:718: UserWarning: Named tensors and all their associated APIs are an experimental feature and subject to change. Please do not use them for anything important until they are released as stable. (Triggered internally at  /pytorch/c10/core/TensorImpl.h:1156.)
  return torch.max_pool2d(input, kernel_size, stride, padding, dilation, ceil_mode)
Phase: train Epoch: 1/1 Loss: 0.6993 Acc: 0.5246        
Phase: validation   Epoch: 1/1 Loss: 0.6432 Acc: 0.6536        
Training completed in 0m 38s
Best test loss: 0.6432 | Best test accuracy: 0.6536
python ./tutorial_quantum_transfer_learning.py  45.20s user 3.36s system 63% cpu 1:16.84 total
```

I can confirm the GPU was used during this demo also as the memory for above process was allocated and freed. Happy to run more if needed.

**Review by albi3ro — APPROVED** [review]
Thanks @mlxd  for checking that it executes :)

Thanks for figuring this out :)

**Pull request merged**
